_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,353 @@
+I believe this is what people call an invariant. A requirement that must hold for the data to have meaning.  
+How to enforce then this constraint ?  
+Let us first define the data types relevant to the problem:  
+
+
+```scala
+sealed trait Seat
+
+case class Seated(row: Int, column: Int) extends Seat
+case class Tribune(tribuneId: String) extends Seat
+case object Standing extends Seat
+```
+
+In what follows it will be useful to discern the type relation above on the following diagram.
+
+<img class=" size-full wp-image-397 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/12/typehiearchy1.png" alt="ActionBuilder" />  
+
+We also have:  
+
+```scala
+case class Ticket[SeatType <: Seat](
+  verificationCode: String,
+  price: Int,
+  seat: SeatType
+)
+```
+
+```scala
+case class Reservation[SeatType <: Seat](
+  userTicket: Ticket[SeatType],
+  guestTicket: Ticket[SeatType]
+)
+```
+
+
+The first thing to notice is that the above does not deliver what we want. The following compiles:  
+
+```scala
+val reservation = Reservation(
+  usertTicket = Ticket("id1", 123, VipTribuneSeat("east")),
+  guestTicket = Ticket("id2", 123, Standing)
+)
+```
+
+The compiler tries to infer the type <code>SeatType</code>for reservation. From the user ticket on line 2 it thinks it is <code>VipTribuneSeat</code>. The compiler is satisfied since this a sub-type of <code>Seat</code> which is an upper type bound on Reservation. This is contradicted on the 3rd line, as we encounter type <code>Standing</code> for the guest ticket. Bad news since type <code>SeatType</code> only line 2 and <code>SeatType</code> on line 3 refer, well, to the same type!  
+Before launching an error, the compiler realizes both <code>Standing </code> and <code>VipTribuneSeat</code> are actually of sub-types of <code>Seat</code> ; this means they can pretend to be their super-type (<code>Seat</code>). Because <code>Seat</code> respects the type bound <code>Seat</code> <:  <code>Seat</code>, the compilation is successful with val reservation: Reservation[Seat].
+
+<strong> There are at least 4 ways to enforce the restriction of the opera-house: </strong>
+
+
+1. using require(condition) on the class constructor of Reservation  
+2. Hiding the public constructor of the class, defining a public one that validates and returns Option[Reservation]
+3. Using generalized type constrains.
+4. Using OOP - subtyping
+
+
+# Using require(condition) on the class constructor of Reservation
+
+```scala
+case class Reservation[SeatInfo <: Seat](
+  userTicket: Ticket[SeatInfo],
+  guestTicket: Ticket[SeatInfo]
+){
+
+  private def verifyTypeRestriction: Boolean = (userTicket.seat, guestTicket.seat) match {
+    case (NormalSeat(_, _), NormalSeat(_, _)) => true
+    case (VipTribuneSeat(_), VipTribuneSeat(_)) => true
+    case (_: Standing.type , _: Standing.type ) => true
+    case _ => false
```

Comment:
I kinda agree.
For some reason I get an error when I do:
` case (Standing, Standing) => ......`
Did not try to understand why, just changed and moved on.

Reply:
You need to add the back ticks (`) otherwise it will try to unapply with the variable named Standing. If you copy the bit on the comment above it should work.